astropy/wcs/utils.py

```diff
@@ -118,6 +140,60 @@ def _wcs_to_celestial_frame_builtin(wcs):
                 representation_type=SphericalRepresentation,
                 obstime=wcs.wcs.dateobs or None,
             )
+        elif (
+            "LN" in xcoord
+            and "LT" in ycoord
+            and "H" not in xcoord
+            and "CR" not in xcoord
+        ):
+            # Coordinates on a planetary body, as defined in
+            # https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2018EA000388
+
+            object_name = SOLAR_SYSTEM_OBJ_DICT.get(xcoord[:2])
+            if object_name is None:
+                raise KeyError(f"unknown solar system object abbreviation {xcoord[:2]}")
+
+            a_radius = wcs.wcs.aux.a_radius
+            b_radius = wcs.wcs.aux.b_radius
+            c_radius = wcs.wcs.aux.c_radius
+            if "bodycentric" in wcs.wcs.name.lower():
+                baserepresentation = BaseBodycentricRepresentation
+                representation_type_name = "BodycentricRepresentation"
+            else:
+                baserepresentation = BaseGeodeticRepresentation
+                representation_type_name = "GeodeticRepresentation"
+
+            if a_radius == b_radius:
+                equatorial_radius = a_radius * u.m
+                flattening = (a_radius - c_radius) / a_radius
+            else:
+                raise NotImplementedError(
+                    "triaxial systems are not supported at this time."
+                )
+
+            # create a new representation class
+            representation_type = solar_system_body_representation_type.get(xcoord[:2])
+            if representation_type is None:
+                representation_type = type(
+                    f"{object_name}{representation_type_name}",
+                    (baserepresentation,),
+                    dict(_equatorial_radius=equatorial_radius, _flattening=flattening),
+                )
+                solar_system_body_representation_type[xcoord[:2]] = representation_type
+
+            # create a new frame class
+            frame = solar_system_body_frames.get(xcoord[:2])
```

Comment:
Same as above: this should be a different frame any time there is a new `object_name` *or* `representation_type`, so again something like
```
@lru_cache(maxsize=100)
def solar_system_body_frame(object_name, representation_type):
    return type(
        f"{object_name}Frame",
        (BaseCoordinateFrame,),
        dict(...),
    )
```